File: Briscola.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <iomanip>
#include <limits>
#define DIM 40

using namespace std;

class Carta{
	private:
		string seme;
		short valore;
		short punteggio;
		bool briscola;
		bool lanciata;
	public:
		Carta(string s, short v, short p, bool b, bool l): seme(s), valore(v), punteggio(p), briscola(b), lanciata(l){}
		
		string& getSeme(){return seme;}
		short& getValore(){return valore;}
		short& getPunteggio(){return punteggio;}
		bool& getBriscola(){return briscola;}
		bool& isLanciata(){return lanciata;}
		
		string name();
		
		ostream& print(ostream& os){
			int i=1,j=4;
			if(this->valore<10) j=5;
			if(this->seme=="Oro") i=3;
			os << "_______" << endl;
			os << "|" << seme <<  setw(i) << "|" << endl;
			os << "|" << valore << setw(j) << "|" << endl;
			os << "|     |" << endl;
			os << "|_____|" << endl;
			return os;
		}
};

ostream& operator<<(ostream& os, Carta* c){
	return c->print(os);
}

string Carta::name(){
	switch(this->getValore()){
		case 1: return "un gran bell'asso";
		case 2: return "un 2";
		case 3: return "un 3";
		case 4: return "un 4";
		case 5: return "un bel 5";
		case 6: return "un bel 6";
		case 7: return "un bel 7";
		case 8: return "una bella donna";
		case 9: return "un bel cavallo";
		case 10: return "un bel re";
	}
}

short setPunteggio(short n){
	switch(n){
		case 1: return 11;
		case 2:
		case 4:
		case 5:
		case 6:
		case 7: return 0;
		case 8: return 2;
		case 3: return 10;
		case 9: return 3;
		case 10: return 4;
	}
}

void init(Carta** vec){
	string a[4]={"Oro","Spade","Coppe","Mazze"};
	for(int i=0; i<4; i++){
		for(int j=0; j<10; j++){
			vec[10*i+j] = new Carta(a[i],j+1,setPunteggio(j+1),false,false);
		}
	}
	cout << "BRISCOLA" << endl << endl;
	cout << "Benvenuto, il mazziere ha preso il mazzo e lo sta mischiando" << endl;
}

void swap(Carta** vec, int i){
	int pos = rand()%DIM;
	Carta aux = *vec[i];
	*vec[i] = *vec[pos];
	*vec[pos] = aux;
}

void mischia(Carta** vec){
	for(int i=0; i<DIM; i++) swap(vec,i);
}

void distribuisci(Carta** vec1, Carta** vec2, int& n1, int n2){
	for(int i=0; i<n2; i++)
		vec2[i] = vec1[i+n1];
	n1+=n2;
}

string setBriscola(Carta** vec){
	string br = vec[39]->getSeme();
	for(int i=0; i<DIM; i++)
		if(vec[i]->getSeme()==br) vec[i]->getBriscola()=true;
	cout << "Il seme di briscola e' " << br << endl;
	system("pause");
	system("cls");
	return br;
}

void noCartaTerra(){
	cout << "_______" << endl
		 << "|     |" << endl
		 << "|     |" << endl
		 << "|     |" << endl
		 << "|_____|" << endl;;
}

void mostra(Carta** gioc, Carta** terra, bool vuoto, string br, int index, int g, int a){
	system("cls");
	cout << "Il seme di briscola e' " << br << setw(br=="Oro" ? 22 : 20) << "MAZZO" << endl;
	cout << "___________________________" << setw(33) << "Carte rimanenti: " << DIM-index << endl;
	cout << "Queste sono le tue carte" << endl;
	for(int i=0; i<3; i++){
		if(gioc[i]->isLanciata()) noCartaTerra();
		else cout << gioc[i];
	}
	cout << "___________________________" << endl;
	cout << "Carta a terra" << endl;
	if(vuoto) noCartaTerra();
	else cout << terra[0];
}

void lancio(Carta** orig, Carta** dest, int i){
	dest[0] = orig[i];
	orig[i]->isLanciata() = true;
}

string initPartita(Carta** mazzo, Carta** gioc, Carta** avv, Carta** terra, int& n1, int n2){
	init(mazzo);
	mischia(mazzo);
	distribuisci(mazzo,gioc,n1,n2);
	distribuisci(mazzo,avv,n1,n2);
	string br = setBriscola(mazzo);
	return br;
}

int match(Carta** vec, Carta** terra, int i){
	int ret = vec[i]->getPunteggio()+terra[0]->getPunteggio();
	if(terra[0]->getBriscola()){
		if(vec[i]->getBriscola()){
			if(vec[i]->getPunteggio()==0 && terra[0]->getPunteggio()==0)
				return vec[i]->getValore()>terra[0]->getValore() ? ret : -ret;
			return vec[i]->getPunteggio()>terra[0]->getPunteggio() ? ret : -ret;
		}
		else return -ret;
	}
	else{
		if(vec[i]->getBriscola()) return ret;
		else{
			if(vec[i]->getSeme()==terra[0]->getSeme()){
				if(vec[i]->getPunteggio()==0 && terra[0]->getPunteggio()==0)
					return vec[i]->getValore()>terra[0]->getValore() ? ret : -ret;
				return vec[i]->getPunteggio()>terra[0]->getPunteggio() ? ret : -ret;
			}
			else return -ret;
		}
	}
}

bool matchL(Carta** vec, Carta** terra, int i){
	if(terra[0]->getBriscola()){
		if(vec[i]->getBriscola()){
			return terra[0]->getValore()>vec[i]->getValore() ? false : true;
		}
	}
	else{
		if(vec[i]->getBriscola()) return true;
		else return terra[0]->getValore()>vec[i]->getValore() ? false : true;
	}
}

void occhiata(Carta** avv){
	system("cls");
	cout << "Minchia mbare, belle carte..." << endl;
	cout << "___________________________" << endl;
	cout << "Queste NON sono le tue carte" << endl;
	for(int i=0; i<3; i++){
		if(avv[i]->isLanciata()) noCartaTerra();
		else cout << avv[i];
	}
	cout << endl;
	system("pause");
}

// gioc sono le carte del giocatore
// avv sono le carte dell'avversario
// terra è la carta a terra (se c'è)
// g è il punteggio del giocatore
// a è il punteggio dell'avversario
// turno dice chi inizia il turno ogni volta (1->giocatore, 0->avversario)
// vuoto dice se a terra non c'è nessuna carta
// index è l'indice del mazzo
// z è l'indice della carta scelta dal giocatore
// q è l'indice della carta scelta dall'avversario
// br è la briscola

bool game(Carta** gioc, Carta** avv, Carta** terra, int& g, int& a, bool& turno, bool& vuoto, int& index, short& z, short& q, string br){
	mostra(gioc,terra,vuoto,br,index,g,a);
	// comincio io
	if(turno){
		cout << endl << "Scegli la carta da lanciare (1) (2) (3) --> ";
		cin >> z;
		if(z==1234){
			occhiata(avv);
			mostra(gioc,terra,vuoto,br,index,g,a);
		}
		while(cin.fail() || z>3 || z<1 || gioc[z-1]->isLanciata()){
			cin.clear();
			cin.ignore(numeric_limits<streamsize>::max(), '\n');
			cout << endl << "Scegli la carta da lanciare (1) (2) (3) --> ";
			cin >> z;
		}
		z--;
		lancio(gioc,terra,z);
		vuoto = false;
		mostra(gioc,terra,vuoto,br,index,g,a);
		cout << endl << "Ora tocca all'avversario" << endl;
		system("pause");
		q = rand()%3;
		mostra(gioc,terra,vuoto,br,index,g,a);
		while(avv[q]->isLanciata()) q = rand()%3;
		if(avv[q]->getPunteggio()==0){
			cout << endl << "Il tuo avversario ci va di liscio con " << avv[q]->name() << (avv[q]->getSeme()=="Oro" ? " d'" : " di ") << avv[q]->getSeme() << endl;
			if(avv[q]->getBriscola()) cout << "Ma tena accura, ittau na briscola" << endl;
		}
		else
			cout << endl << "Il tuo avversario sta per lanciare " << avv[q]->name() << (avv[q]->getSeme()=="Oro" ? " d'" : " di ") << avv[q]->getSeme() << endl;
		avv[q]->isLanciata() = true;
		system("pause");
		if(match(avv,terra,q)>0){
			a+=match(avv,terra,q);
			cout << endl << "Ha preso l'avversario" << endl;
			system("pause");
			vuoto = true;
			turno = false;
			return true;
		}
		else if(match(avv,terra,q)<0){
			g-=match(avv,terra,q);
			cout << endl << "Hai preso tu" << endl;
			system("pause");
			vuoto = true;
			turno = true;
			return true;
		}
		else{
			if(matchL(avv,terra,q)){
				cout << endl << "Ha preso l'avversario" << endl;
				system("pause");
				vuoto = true;
				turno = false;
				return true;
			}
			else{
				cout << endl << "Hai preso tu" << endl;
				system("pause");
				vuoto = true;
				turno = true;
				return true;
			}
		}
	}
	// comincia l'avversario
	else{
		cout << endl << "Tocca all'avversario" << endl;
		q = rand()%3;
		while(avv[q]->isLanciata()) q = rand()%3;
		lancio(avv,terra,q);
		vuoto = false;
		system("pause");
		mostra(gioc,terra,vuoto,br,index,g,a);
		if(avv[q]->getPunteggio()==0) cout << endl << "Ci va di liscio" << endl;
		else if(avv[q]->getPunteggio()>=10) cout << endl << "Mizzica, ittau nu bellu caricu" << endl;
		cout << endl << "Ora tocca a te" << endl;
		system("pause");
		mostra(gioc,terra,vuoto,br,index,g,a);
		cout << endl << "Scegli la carta da lanciare (1) (2) (3) --> ";
		cin >> z;
		while(cin.fail() || z>3 || z<1 || gioc[z-1]->isLanciata()){
			cin.clear();
			cin.ignore(numeric_limits<streamsize>::max(), '\n');
			cout << endl << "Scegli la carta da lanciare (1) (2) (3) --> ";
			cin >> z;
		}
		z--;
		gioc[z]->isLanciata() = true;
		if(match(gioc,terra,z)>0){
			g+=match(gioc,terra,z);
			cout << endl << "Hai preso tu" << endl;
			system("pause");
			vuoto = true;
			turno = true;
			return true;
		}
		else if(match(gioc,terra,z)<0){
			a-=match(gioc,terra,z);
			cout << endl << "Ha preso l'avversario" << endl;
			system("pause");
			vuoto = true;
			turno = false;
			return true;
		}
		else{
			if(matchL(gioc,terra,z)){
				cout << endl << "Hai preso tu" << endl;
				system("pause");
				vuoto = true;
				turno = true;
				return true;
			}
			else{
				cout << endl << "Ha preso l'avversario" << endl;
				system("pause");
				vuoto = true;
				turno = false;
				return true;
			}
		}
	}
}

void pesca(Carta** mazzo, Carta** vec, short n, int& index){
	vec[n] = mazzo[index++];
}


int main(){
	srand(time(0));
	Carta* mazzo[DIM];
	Carta* gioc[3];
	Carta* avv[3];
	Carta* terra[1];
	int index = 0, punteggioGioc = 0, punteggioAvv = 0;
	bool turno = true, vuoto = true, g = true;
	short z,q;
	
	string br = initPartita(mazzo,gioc,avv,terra,index,3);
	
	while(g){
		g = game(gioc,avv,terra,punteggioGioc,punteggioAvv,turno,vuoto,index,z,q,br);
		pesca(mazzo,gioc,z,index);
		pesca(mazzo,avv,q,index);
		if(index>39){
			cout << endl << "ATTENZIONE: e' appena finito il mazzo, giocatela bene ora" << endl;
			system("pause");
			g = false;
		}
	}
	
	for(int i=0; i<3; i++){
		game(gioc,avv,terra,punteggioGioc,punteggioAvv,turno,vuoto,index,z,q,br);
	}
	
	mostra(gioc,terra,vuoto,br,index,punteggioGioc,punteggioAvv);
	cout << endl << "La partita e' giunta alla fine, ed il vincitore e'..." << endl;
	if(punteggioGioc==punteggioAvv) cout << "Wow, non me l'aspettavo, questo e' un bel pareggio" << endl;
	else{
		cout << (punteggioGioc>punteggioAvv ? "TU!!! Grandissimo, hai vinto con un bel punteggio di " : "Ehm, non sei tu, mi spiace ma hai perso, il tuo avversario ha totalizato un punteggio di ");
		cout << (punteggioGioc>punteggioAvv ? punteggioGioc : punteggioAvv) << " punti" << endl;
	}
	system("pause");
	
	return 0;
}

File: Briscola.py
from os import system
from random import randint

# variable definitions area:
# arrays
deck, player, opponent, ground = [], [], [], [0]
# int
index, player_points, opponent_points, ind1, ind2 = 0, 0, 0, 0, 0
# booleans
turn, no_ground, game = True, True, True

# constant value
DIM = 40

# class: it is the class which defines a card
class Card:
    # constructor
    def __init__(self, seed: str, value: int, points: int, briscola: bool, launched: bool):
        self.seed = seed
        self.value = value
        self.points = points
        self.briscola = briscola
        self.launched = launched
    # function: gives to each card a name
    def name(self) -> str:
        match self.value:
            case 1: return "un gran bell'asso"
            case 2: return "un 2"
            case 3: return "un 3"
            case 4: return "un 4"
            case 5: return "un bel 5"
            case 6: return "un bel 6"
            case 7: return "un bel 7"
            case 8: return "una bella donna"
            case 9: return "un bel cavallo"
            case 10: return "un bel re"

# function: defines the way to print the cards
def print_card(obj :Card) -> None:
    print("_________")
    if obj.seed == "Oro":
        print("|" + obj.seed + "    |")
    else:
        print("|" + obj.seed + "  |")
    match obj.value:
        case 1:
            print("|Asso   |")
        case 8:
            print("|Donna  |")
        case 9:
            print("|Cavallo|")
        case 10:
           print("|Re     |") 
        case _:
            print("|" + str(obj.value) + "      |")
    print("|       |")
    print("|       |")
    print("|_______|")

# function: sets the amount of points for each card
def setPoints(n: int) -> int:
    match n:
        case 1: return 11
        case 2,4,5,6,7: return 0
        case 8: return 2
        case 9: return 3
        case 10: return 4
        case 3: return 10

# function: initializes the deck
def init(list: Card) -> None:
    a = ["Oro","Spade","Coppe","Mazze"]
    for i in range(4):
        for j in range(10):
            list.append(Card(a[i],j+1,setPoints(j+1),False,False))
    print("BRISCOLA")
    print("Benvenuto, il mazziere ha preso il mazzo e lo sta mischiando")

# function: swap two cards position
def swap(list: Card, i: int) -> None:
    pos = randint(0,DIM-1)
    aux = list[i]
    list[i] = list[pos]
    list[pos] = aux

# function: mix the deck
def mix(list: Card) -> None:
    for i in range(DIM): swap(list,i)
    
# function: gives the three first cards to each player
def distribute(deck: Card, list: Card) -> None:
    global index
    for i in range(3): list.append(deck[i+index])
    index += 3

# function: set the briscola status
def setBriscola(list: Card) -> str:
    br = list[39].seed
    for i in range(DIM):
        if(list[i].seed == br): list[i].briscola = True
    print("Il seme di briscola e' " + br)
    system("pause")
    system("cls")
    return br

# function: defines the way to print a ground without any card
def no_card() -> None:
    print("_________")
    print("|       |")
    print("|       |")
    print("|       |")
    print("|       |")
    print("|_______|")

# function: shows the playground
def show(pl: Card, gr: Card, noc: bool, br: str) -> None:
    system("cls")
    print("Il seme di briscola e' " + br + ("            MAZZO" if br == "Oro" else "          MAZZO"))
    print("___________________________" + "           Carte rimanenti: " + str(DIM-index))
    print("Queste sono le tue carte")
    for i in range(3):
        if(pl[i].launched): no_card()
        else: print_card(pl[i])
    print("___________________________")
    print("Carta a terra")
    if(noc): no_card()
    else: print_card(gr[0])

# function: executes the launch of a card on the playground
def launch(orig: Card, dest: Card, i: int):
    dest[0] = orig[i]
    orig[i].launched = True

# function: initializes the game
def initGame(deck: Card, pl: Card, op: Card) -> str:
    init(deck)
    mix(deck)
    distribute(deck,pl)
    distribute(deck,op)
    br = setBriscola(deck)
    return br

# function: draws a card from the deck
def draw(deck: Card, list: Card, n: int) -> None:
    global index
    list[n] = deck[index]
    index += 1

# function: choose who win a match
def match(list: Card, ground: Card, i: int) -> int:
    first = list[i].points if list[i].points is not None else 0
    second = ground[0].points if ground[0].points is not None else 0
    ret = first + second
    if(ground[0].briscola):
        if(list[i].briscola):
            return (ret if first > second else -ret)
        else: return -ret
    else:
        if(list[i].briscola): return ret
        else:
            if(list[i].seed == ground[0].seed):
                return (ret if first > second else -ret)   
            else: return -ret

# function: choose who win a match with "liscie" cards
def matchL(list: Card, ground: Card, i: int) -> bool:
    if(ground[0].briscola):
        if(list[i].briscola):
            return (True if list[i].value > ground[0].value else False)
    else:
        if(list[i].briscola): return True
        else: return (True if list[i].value > ground[0].value else False)

# function: it's not really a legitimate thing
def occhiata(opp: Card):
    system("cls")
    print("Minchia mbare, belle carte...")
    print("___________________________")
    print("Queste NON sono le tue carte")
    for i in range(3):
        if(opp[i].launched): no_card()
        else: print_card(opp[i])
    print('\n')
    system("pause")

# function: this is the most important function of the program
def engine(pl: Card, op: Card, gr: Card, br: str) -> bool:
    global ind1, ind2, no_ground, turn, player_points, opponent_points
    show(pl,gr,no_ground,br)
    if(turn):
        ind1 = input("\nScegli la carta da lanciare (1) (2) (3) --> ")
        while(str(ind1) == '' or not str(ind1).isdigit() or int(ind1)>3 or int(ind1)<1 or pl[int(ind1)-1].launched):
            if(str(ind1).isdigit() and int(ind1) == 1234):
                occhiata(op)
                show(pl,gr,no_ground,br)
            ind1 = input("\nScegli la carta da lanciare (1) (2) (3) --> ")
        ind1 = int(ind1)
        ind1 -= 1
        launch(pl,gr,ind1)
        no_ground = False
        show(pl,gr,no_ground,br)
        print("\nOra tocca all'avversario")
        system("pause")
        show(pl,gr,no_ground,br)
        ind2 = randint(0,2)
        while(op[ind2].launched): ind2 = randint(0,2)
        if(op[ind2].points is None):
            print("\nIl tuo avversario ci va di liscio con " + op[ind2].name() + (" d'" if op[ind2].seed == "Oro" else " di ") + op[ind2].seed)
            if(op[ind2].briscola): print("Ma fai attenzione, ha lanciato una briscola")
        else:
            print("\nIl tuo avversario sta per lanciare " + op[ind2].name() + (" d'" if op[ind2].seed == "Oro" else " di ") + op[ind2].seed)
        op[ind2].launched = True
        system("pause")
        pt = match(op,gr,ind2)
        if(pt > 0):
            opponent_points += pt
            print("\nHa preso l'avversario")
            system("pause")
            no_ground = True
            turn = False
            return True
        elif(pt < 0):
            player_points -= pt
            print("\nHai preso tu")
            system("pause")
            no_ground = True
            turn = True
            return True
        else:
            if(matchL(op,gr,ind2)):
                print("\nHa preso l'avversario")
                system("pause")
                no_ground = True
                turn = False
                return True
            else:
                print("\nHai preso tu")
                system("pause")
                no_ground = True
                turn = True
                return True
    else:
        print("\nTocca all'avversario")
        ind2 = randint(0,2)
        while(op[ind2].launched): ind2 = randint(0,2)
        launch(op,gr,ind2)
        no_ground = False
        system("pause")
        show(pl,gr,no_ground,br)
        if(op[ind2].points is None): print("\nCi va di liscio")
        elif(op[ind2].points >= 10): print("\nIl tuo avversario ha buttato un bel carico")
        print("\nOra tocca a te")
        system("pause")
        show(pl,gr,no_ground,br)
        ind1 = input("\nScegli la carta da lanciare (1) (2) (3) --> ")
        while(str(ind1) == '' or not str(ind1).isdigit() or int(ind1)>3 or int(ind1)<1 or pl[int(ind1)-1].launched):
            ind1 = input("\nScegli la carta da lanciare (1) (2) (3) --> ")
        ind1 = int(ind1)
        ind1 -= 1
        pl[ind1].launched = True
        pt = match(pl,gr,ind1)
        if(pt > 0):
            player_points += pt
            print("\nHai preso tu")
            system("pause")
            no_ground = True
            turn = True
            return True
        elif(pt < 0):
            opponent_points -= pt
            print("\nHa preso l'avversario")
            system("pause")
            no_ground = True
            turn = False
            return True
        else:
            if(matchL(pl,gr,ind1)):
                print("\nHai preso tu")
                system("pause")
                no_ground = True
                turn = True
                return True
            else:
                print("\nHa preso l'avversario")
                system("pause")
                no_ground = True
                turn = False
                return True            

# main function: here starts the execution
def main() -> None:
    global game,no_ground 
    br = initGame(deck,player,opponent)
    while(game):
        game = engine(player,opponent,ground,br)
        draw(deck,player,ind1)
        draw(deck,opponent,ind2)
        if(index > 39):
            show(player,ground,no_ground,br)
            print("\n!!! ATTENZIONE !!!\nIl mazzo e' appena finito, giocatela bene ora")
            system("pause")
            game = False
    for i in range(3): engine(player,opponent,ground,br)
    show(player,ground,no_ground,br)
    print("\nLa partita e' giunta alla fine, ed il vincitore e'...")
    system("pause")
    if(player_points == opponent_points): print("Wow, non me l'aspettavo, questo e' un bel pareggio")
    else:
        print("TU!!! Grandissimo, hai vinto con un bel punteggio di: " if player_points > opponent_points else "Ehm, non sei tu, mi spiace ma hai perso, il tuo avversario ha totalizato un punteggio di:", end = "\n\n")
        print("--> " + (str(player_points) if player_points > opponent_points else str(opponent_points)) + " <-- " + " punti")
    system("pause");

if __name__ == "__main__": main()